system/Debug/Toolbar.php

```diff
@@ -86,55 +82,226 @@ public function __construct(BaseConfig $config)
-		$segmentDuration = $this->roundTo($totalTime/7, 5);
-		$segmentCount    = (int)ceil($totalTime/$segmentDuration);
-		$varData         = $this->collectVarData();
+		foreach ($this->collectVarData() as $heading => $items)
+		{
+			$vardata = [];
 
-		ob_start();
-		include(__DIR__.'/Toolbar/Views/toolbar.tpl.php');
-		$output = ob_get_contents();
-		ob_end_clean();
+			if (is_array($items))
+			{
+				foreach ($items as $key => $value)
+				{
+					$vardata[esc($key)] = is_string($value) ? esc($value) : print_r($value, true);
+				}
+			}
 
-		return $output;
+			$data['vars']['varData'][esc($heading)] = $vardata;
+		}
+
+		if (isset($_SESSION) && ! empty($_SESSION))
+		{
+			foreach ($_SESSION as $key => $value)
+			{
+				$data['vars']['session'][esc($key)] = is_string($value) ? esc($value) : print_r($value, true);
+			}
+		}
+
+		foreach ($request->getGet() as $name => $value)
+		{
+			$data['vars']['get'][esc($name)] = esc($value);
+		}
+
+		foreach ($request->getPost() as $name => $value)
+		{
+			$data['vars']['post'][esc($name)] = is_array($value) ? esc(print_r($value, true)) : esc($value);
+		}
+
+		foreach ($request->getHeaders() as $header => $value)
+		{
+			if (empty($value))
+			{
+				continue;
+			}
+
+			if (! is_array($value))
+			{
+				$value = [$value];
+			}
+
+			foreach ($value as $h)
+			{
+				$data['vars']['headers'][esc($h->getName())] = esc($h->getValueLine());
+			}
+		}
+
+		foreach ($request->getCookie() as $name => $value)
+		{
+			$data['vars']['cookies'][esc($name)] = esc($value);
+		}
+
+		$data['vars']['request'] = ($request->isSecure() ? 'HTTPS' : 'HTTP').'/'.$request->getProtocolVersion();
+
+		$data['vars']['response'] = [
+			'statusCode'      => $response->getStatusCode(),
+			'reason'          => esc($response->getReason()),
+			'contentType'     => esc($response->getHeaderLine('content-type')),
+		];
+
+		$data['config'] = \CodeIgniter\Debug\Toolbar\Collectors\Config::display();
+
+		return json_encode($data);
 	}
 
 	//--------------------------------------------------------------------
 
 	/**
-	 * Called within the view to display the timeline itself.
+	 * Format output
 	 *
-	 * @param int $segmentCount
-	 * @param int $segmentDuration
+	 * @param  string $data   JSON encoded Toolbar data
+	 * @param  string $format html, json, xml
 	 *
 	 * @return string
 	 */
-	protected function renderTimeline(int $segmentCount, int $segmentDuration): string
+	protected static function format(string $data, string $format = 'html')
 	{
-		$displayTime = $segmentCount*$segmentDuration;
+		$data = json_decode($data, true);
+
+		// History must be loaded on the fly
+		$filenames = glob(WRITEPATH.'debugbar/debugbar_*');
+		$total     = count($filenames);
+		rsort($filenames);
 
-		$rows = $this->collectTimelineData();
+		$files = [];
+
+		$current = Services::request()->getGet('debugbar_time');
+
+		for ($i = 0; $i < $total; $i++)
+		{
+			// Oldest files will be deleted
+			if ($i >= 30)
```

Comment:
Please make the value here set in Config\App. Some people may have semi-sensitive information they don't want kept around long on their servers during testing, so they should be able to define the amount to keep around.